descqa/NumberDensityVersusRedshift.py

```diff
@@ -0,0 +1,307 @@
+
+        #get catalog data
+        mag_field = galaxy_catalog.first_available(*self.possible_mag_fields)
+        if not mag_field:
+            return TestResult(skipped = True,summary = 'Missing requested quantities')
+        catalog_data = self.get_catalog_data(galaxy_catalog, [self.z, mag_field], filters=self.filters)
+        filtername = mag_field.partition(self.band + '_')[-1].upper()
+        filelabel = '_'.join([filtername,self.band])
+
+        fig, ax  =  plt.subplots(self.nrows,self.ncolumns,figsize=(figx_p,figy_p),sharex='col')
+        self.yaxis = 'P(z|m)' if self.normed else 'N(z|m)'
+
+        catalog_color = next(self.colors)
+        
+        #loop over magnitude cuts and make plots
+        results = {}
+        for n,cut in enumerate(self.mag_lo):
+            #initialize default values to use for validation-data fits
+            z0 = 0.
+            z0err = 0.
+            #get masks for magnitude cuts and get validation-data fits depending on selected observation
+            if self.observation.find(Coil2004)!= -1:
+                mask = (catalog_data[mag_field]<=cut) & (catalog_data[mag_field]>self.mag_hi[n])
+                cutlabel = ' '.join(['$',str(self.mag_hi[n]),'< $',self.band,'$<=',str(cut),'$'])
+                
+                #find correct row in validation data fits to use in plot
+                omask = (self.mag_hi.astype(int)==int(self.mag_hi[n])) & (self.mag_lo.astype(int)==int(cut))
+                z0 = self.validation_data['z0values'][omask]
+                if('z0errors' in self.validation_data.keys()):
+                    z0err = self.validation_data['z0errors'][omask]
+            else:
+                mask = (catalog_data[mag_field]<cut)
+                cutlabel =  ' '.join([self.band,'$<',str(cut),'$'])
+                if(self.observation.find(DEEP2_JAN)!= -1):
+                    z0 = self.validation_data['z0const'] + self.validation_data['z0linear']*(cut)
+            total = '(# of galaxies = {})'.format(np.sum(mask))
+ 
+            #bin catalog_data
+            N,binEdges = np.histogram(catalog_data[self.z][mask],bins=self.zbins)
+            sumz,binEdges = np.histogram(catalog_data[self.z][mask],bins=self.zbins,weights=catalog_data[self.z][mask])
+            meanz = sumz/N
+
+            #make subplot
+            ncol = int(n%self.ncolumns)
+            nrow = n//self.ncolumns
+            catalog_label = ' '.join([catalog_name,re.sub(self.band,filtername+' '+self.band,cutlabel)])
+            validation_label = ' '.join([self.validation_data['label'],cutlabel])
+            reskey = re.sub('\$','',cutlabel)
+            results[reskey] = self.make_subplot(meanz,catalog_data[self.z][mask],n,ax[nrow,ncol],z0,z0err,catalog_color,validation_color,catalog_label,validation_label)
+            results[reskey]['total'] = total
+
+            #add curve for this catalog to summary plot
+            if self.first_pass:  #add validation data if evaluating first catalog
```

Comment:
Instead of checking self.first_pass, moving this part to `__init__` (i.e., load and plot validation data in `__init__`)